docs/source/using_websockets.rst

```diff
@@ -4,30 +4,186 @@ Using WebSockets in Dallinger Experiments
+a JSON object. When an experiment receives a message where the payload is a JSON
+object containing either a `node_id` property or a `participant_id` value (the
+properties `participant_id`, `sender`, and `client.participant_id` are treated
+as equivalent), then the experiment will queue such messages to be processed
+asynchronously. Otherwise, the messages will be processed synchronously by an
+experiment instance. In either case, the
+:func:`~dallinger.experiment.Experiment.receive_message` method will handle the
+message.
+
+Client Implementation
+---------------------
+
+The default experiment layout includes a `basic websocket communication library
+<https://www.npmjs.com/package/reconnecting-websocket>`_ which implements a
+`ReconnectingWebSocket` object that can be used to establish channel
+subscriptions, send messages to various channels, and receive messages on
+subscribed channels.
+
+Typically experiments set up a WebSocket connection after completing the initial
+call to `createAgent` using code similar to this
+
+::
+    var broadcast_socket;
+
+    var open_socket = function (channel_id) {
+        var ws_scheme = (window.location.protocol === "https:") ? 'wss://' : 'ws://';
+        // Setup a websocket connection to the channel, passing our worker_id and participant_id
+        socket = new ReconnectingWebSocket(
+            ws_scheme + location.host + "/chat?channel=" + channel_id +"&worker_id=" + dallinger.identity.workerId + '&participant_id=' + dallinger.identity.participantId
+        );
+        // Once the connection is established, send an initial message to the channel
+        socket.onopen(function () {
+            socket.send(channel_id + ':{"message": "Hello world!"}');
+        });
+        // Handle any incoming messages
+        socket.onmessage = function (msg) {
+            // Ignore messages not from the channel subscribed channel
+            if (msg.data.indexOf(channel_id + ':') !== 0) { return; }
+            // Parse the payload
+            var data = JSON.parse(msg.data.substring(channel_id.length + 1));
+            // Example message data
+            var type = data.type;
+            // Take different actions based on message type
+            switch(type) {
+               ...
+            }
+        };
+        return socket;
+    };
+
+    // Create the agent.
+    var create_agent = function() {
+        dallinger.createAgent()
+            .done(function (resp) {
+                ...
+                broadcast_socket = open_socket("broadcast_channel");
+            })
+            .fail(function (rejection) {
+                ...
+            });
+    };
+
+
+When establishing a channel subscription using the `/chat` route, the client can
+include `worker_id` and `participant_id` values which will be included in the
+`"dallinger_control"` channel messages alerting the experiment to WebSocket
+connection, disconnection, subscription, and un-subscription events.
```

Comment:
So if, when I initially connect to the websocket on the client side, I include `worker_id` and/or `participant_id` values, they will be automatically forwarded to this other channel somehow? Including these values on messages set _after_ initial setup don't get forwarded in this way?

Reply:
It might be nice to insert these values into all payloads, but we're currently very agnostic about what a message payload contains. The only thing in Dallinger that makes assumptions about the payload is the async stuff discussed above, which only gets triggered for specific types of payloads.